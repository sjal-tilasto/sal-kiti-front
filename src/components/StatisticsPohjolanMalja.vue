<template>
  <div>
    <b-row>
      <b-col>
        <h3 class="bg-sal-orange">
          {{ $tc("statistics.sal.pohjolan_malja", 2) }}
        </h3>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-alert v-if="errors.main" variant="danger" show>
          <ul>
            <li v-for="e in errors.main" v-bind:key="e">{{ e }}</li>
          </ul>
        </b-alert>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-button
          v-for="y in year_range"
          v-bind:key="y"
          variant="light"
          class="btn-orange space-right space-down"
          v-on:click="getStatistics(y)"
          :pressed="y === year"
        >
          {{ y }}
        </b-button>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-table small :fields="resultFields" :items="results.results">
          <template v-slot:cell(value)="data">
            {{ data.item.value }}
          </template>
        </b-table>
        <div v-show="loading">
          <b-spinner label="Loading..."></b-spinner>
        </div>
      </b-col>
    </b-row>
  </div>
</template>

<script>
/**
 * Displays Pohjolan malja statistics
 */
import { HTTP } from "../api/BaseApi.js";
import errorParser from "../utils/ErrorParser";

export default {
  name: "StatisticsPM",
  data() {
    return {
      currentYear: new Date().getFullYear(),
      errors: {},
      loading: false,
      results: [],
      year: new Date().getFullYear()
    };
  },
  computed: {
    /**
     * Sets fields list for the statistics list
     *
     * @returns {array} fields list
     */
    resultFields: function () {
      return [
        {
          key: "organization.name",
          label: this.$t("athlete.club")
        },
        {
          key: "organization.abbreviation",
          label: this.$t("abbreviation")
        },
        { key: "value", label: this.$tc("result.result", 1) }
      ];
    },
    /**
     * Sets year options list
     *
     * @returns {array} year list
     */
    year_range: function () {
      let years = [];
      let startYear = 2020;
      while (startYear <= this.currentYear) {
        years.push(startYear++);
      }
      return years;
    }
  },
  mounted() {
    this.getStatistics(this.year);
  },
  methods: {
    /**
     * Fetch Pohjolan malja statistics for a year from API
     *
     * @param {string} year
     * @returns {Promise<void>}
     */
    async getStatistics(year) {
      this.year = year;
      this.loading = true;
      this.results = [];
      this.$set(this.errors, "main", null);
      HTTP.get("sal/pohjolanmalja/" + year + "/")
        .then((response) => {
          this.results = response.data || [];
        })
        .catch((error) => {
          this.$set(this.errors, "main", errorParser.generic.bind(this)(error));
        })
        .finally(() => (this.loading = false));
    }
  }
};
</script>

<style scoped></style>
