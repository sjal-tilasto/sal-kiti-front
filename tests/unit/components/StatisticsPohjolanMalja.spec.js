import { createLocalVue, mount } from "@vue/test-utils";
import flushPromises from "flush-promises";
import StatisticsPohjolanMalja from "@/components/StatisticsPohjolanMalja.vue";
import BootstrapVue from "bootstrap-vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(BootstrapVue);
localVue.use(Vuex);

describe("StatisticsPohjolanMalja.vue", () => {
  it("renders result list", async () => {
    const wrapper = mount(StatisticsPohjolanMalja, {
      localVue
    });
    wrapper.setData({
      currentYear: 2021,
      year: 2021
    });
    await flushPromises();
    expect(wrapper.element).toMatchSnapshot();
  });
});
