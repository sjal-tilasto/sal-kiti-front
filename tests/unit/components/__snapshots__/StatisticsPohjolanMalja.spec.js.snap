// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPohjolanMalja.vue renders result list 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <h3
        class="bg-sal-orange"
      >
        
        statistics.sal.pohjolan_malja
      
      </h3>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col"
    >
      <!---->
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col"
    >
      <button
        aria-pressed="true"
        autocomplete="off"
        class="btn btn-orange space-right space-down btn-light active"
        type="button"
      >
        
        2020
      
      </button>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col"
    >
      <table
        aria-busy="false"
        aria-colcount="3"
        class="table b-table table-sm"
        id="__BVID__5"
        role="table"
      >
        <!---->
        <!---->
        <thead
          class=""
          role="rowgroup"
        >
          <!---->
          <tr
            class=""
            role="row"
          >
            <th
              aria-colindex="1"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                athlete.club
              </div>
            </th>
            <th
              aria-colindex="2"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                abbreviation
              </div>
            </th>
            <th
              aria-colindex="3"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                result.result
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <!---->
          <tr
            class=""
            role="row"
          >
            <td
              aria-colindex="1"
              class=""
              role="cell"
            >
              Local Club
            </td>
            <td
              aria-colindex="2"
              class=""
              role="cell"
            >
              LC
            </td>
            <td
              aria-colindex="3"
              class=""
              role="cell"
            >
              
          8
        
            </td>
          </tr>
          <tr
            class=""
            role="row"
          >
            <td
              aria-colindex="1"
              class=""
              role="cell"
            >
              Village Club
            </td>
            <td
              aria-colindex="2"
              class=""
              role="cell"
            >
              VC
            </td>
            <td
              aria-colindex="3"
              class=""
              role="cell"
            >
              
          2
        
            </td>
          </tr>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
       
      <div
        style="display: none;"
      >
        <span
          class="spinner-border"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
